File: apps/vscode-extension/src/repo.ts
import * as vscode from "vscode";
import * as path from "path";
import * as fs from "fs";

export interface RepoInfo {
  owner: string;
  repo: string;
  /** Full "owner/repo" string. */
  fullName: string;
}

/** Per-workspace-folder cache of detected repo info. */
const cache = new Map<string, RepoInfo | null>();

/**
 * Detect owner/repo for the first workspace folder (or a specific one).
 * Looks for `.jj/repo/store/git` config first, then falls back to `.git/config`.
 * Results are cached per workspace folder path.
 */
export function detectRepo(
  folder?: vscode.WorkspaceFolder,
): RepoInfo | undefined {
  const ws =
    folder ?? vscode.workspace.workspaceFolders?.[0];
  if (!ws) {
    return undefined;
  }

  const root = ws.uri.fsPath;

  if (cache.has(root)) {
    return cache.get(root) ?? undefined;
  }

  const info = tryJJ(root) ?? tryGit(root);
  cache.set(root, info ?? null);
  return info ?? undefined;
}

/** Clear the cache (useful when workspace folders change). */
export function clearRepoCache(): void {
  cache.clear();
}

// ── Internal helpers ──────────────────────────────────────────────

/** Parse a jjhub.tech remote from jj colocated git config. */
function tryJJ(root: string): RepoInfo | undefined {
  // jj colocated repos store their git repo at .jj/repo/store/git
  // which is often a path reference. Try the colocated .git/config first.
  const colocatedGitConfig = path.join(root, ".git", "config");
  const jjGitDir = path.join(root, ".jj", "repo", "store", "git");

  // Prefer colocated .git/config (most common jj setup)
  if (fs.existsSync(colocatedGitConfig)) {
    return parseGitConfig(colocatedGitConfig);
  }

  // If .jj exists but no colocated git, check if there's a git dir pointer
  if (fs.existsSync(jjGitDir)) {
    try {
      const stat = fs.statSync(jjGitDir);
      if (stat.isDirectory()) {
        const configPath = path.join(jjGitDir, "config");
        if (fs.existsSync(configPath)) {
          return parseGitConfig(configPath);
        }
      } else {
        // It might be a file containing a path to the git dir
        const content = fs.readFileSync(jjGitDir, "utf-8").trim();
        const resolvedConfig = path.resolve(root, content, "config");
        if (fs.existsSync(resolvedConfig)) {
          return parseGitConfig(resolvedConfig);
        }
      }
    } catch {
      // Ignore fs errors
    }
  }

  return undefined;
}

/** Parse remote URL from .git/config. */
function tryGit(root: string): RepoInfo | undefined {
  const configPath = path.join(root, ".git", "config");
  if (!fs.existsSync(configPath)) {
    // .git might be a file (worktree / submodule)
    const gitFile = path.join(root, ".git");
    if (fs.existsSync(gitFile)) {
      try {
        const stat = fs.statSync(gitFile);
        if (!stat.isDirectory()) {
          const content = fs.readFileSync(gitFile, "utf-8").trim();
          const match = content.match(/^gitdir:\s*(.+)$/m);
          if (match) {
            const resolvedConfig = path.resolve(root, match[1], "config");
            if (fs.existsSync(resolvedConfig)) {
              return parseGitConfig(resolvedConfig);
            }
          }
        }
      } catch {
        // Ignore
      }
    }
    return undefined;
  }
  return parseGitConfig(configPath);
}

/**
 * Parse a git config file and extract the first remote URL that points
 * to jjhub.tech, returning the owner/repo.
 */
function parseGitConfig(configPath: string): RepoInfo | undefined {
  try {
    const content = fs.readFileSync(configPath, "utf-8");
    return parseRemoteUrl(content);
  } catch {
    return undefined;
  }
}

/** Extract owner/repo from remote URL lines in git config content. */
function parseRemoteUrl(content: string): RepoInfo | undefined {
  // Match SSH: git@jjhub.tech:owner/repo.git
  // Match SSH: ssh://git@ssh.jjhub.tech/owner/repo.git
  // Match HTTPS: https://jjhub.tech/owner/repo.git
  const urlLines = content.match(/url\s*=\s*(.+)/g);
  if (!urlLines) {
    return undefined;
  }

  for (const line of urlLines) {
    const url = line.replace(/url\s*=\s*/, "").trim();

    // SSH style: git@jjhub.tech:owner/repo.git
    const sshMatch = url.match(
      /git@(?:ssh\.)?jjhub\.tech[:/]([^/]+)\/([^/]+?)(?:\.git)?$/,
    );
    if (sshMatch) {
      return {
        owner: sshMatch[1],
        repo: sshMatch[2],
        fullName: `${sshMatch[1]}/${sshMatch[2]}`,
      };
    }

    // HTTPS style: https://jjhub.tech/owner/repo.git
    const httpsMatch = url.match(
      /https?:\/\/(?:api\.)?jjhub\.tech\/([^/]+)\/([^/]+?)(?:\.git)?$/,
    );
    if (httpsMatch) {
      return {
        owner: httpsMatch[1],
        repo: httpsMatch[2],
        fullName: `${httpsMatch[1]}/${httpsMatch[2]}`,
      };
    }
  }

  // Fallback: try any remote, not just jjhub.tech
  for (const line of urlLines) {
    const url = line.replace(/url\s*=\s*/, "").trim();

    const sshFallback = url.match(
      /git@[^:]+:([^/]+)\/([^/]+?)(?:\.git)?$/,
    );
    if (sshFallback) {
      return {
        owner: sshFallback[1],
        repo: sshFallback[2],
        fullName: `${sshFallback[1]}/${sshFallback[2]}`,
      };
    }

    const httpsFallback = url.match(
      /https?:\/\/[^/]+\/([^/]+)\/([^/]+?)(?:\.git)?$/,
    );
    if (httpsFallback) {
      return {
        owner: httpsFallback[1],
        repo: httpsFallback[2],
        fullName: `${httpsFallback[1]}/${httpsFallback[2]}`,
      };
    }
  }

  return undefined;
}

File: apps/nvim-plugin/lua/jjhub/repo.lua
local M = {}

--- Cache of owner/repo per working directory.
---@type table<string, {owner: string, repo: string}|false>
local cache = {}

--- Parse a remote URL to extract owner/repo.
--- Supports SSH (git@host:owner/repo.git) and HTTPS (https://host/owner/repo.git) formats.
---@param url string
---@return string|nil owner
---@return string|nil repo
local function parse_remote_url(url)
  -- SSH format: git@jjhub.tech:owner/repo.git or ssh://git@host/owner/repo.git
  local owner, repo = url:match("[:/@]([^/]+)/([^/%.]+)%.git%s*$")
  if owner and repo then
    return owner, repo
  end
  -- Without .git suffix
  owner, repo = url:match("[:/@]([^/]+)/([^/%.%s]+)%s*$")
  if owner and repo then
    return owner, repo
  end
  return nil, nil
end

--- Read a git config file and extract the first remote URL that points to jjhub.tech.
--- Falls back to the first remote URL if no jjhub remote is found.
---@param config_path string
---@return string|nil owner
---@return string|nil repo
local function parse_git_config(config_path)
  local f = io.open(config_path, "r")
  if not f then
    return nil, nil
  end
  local content = f:read("*a")
  f:close()

  local first_owner, first_repo = nil, nil

  -- Iterate over remote sections
  for url in content:gmatch('%[remote %"[^"]+"%].-url%s*=%s*([^\n]+)') do
    local owner, repo = parse_remote_url(url)
    if owner and repo then
      -- Prefer jjhub.tech remotes
      if url:find("jjhub%.tech") then
        return owner, repo
      end
      if not first_owner then
        first_owner, first_repo = owner, repo
      end
    end
  end

  return first_owner, first_repo
end

--- Detect owner/repo from the given directory by looking for .jj or .git configs.
---@param cwd string
---@return string|nil owner
---@return string|nil repo
local function detect_from_dir(cwd)
  -- Walk up from cwd to find .jj or .git
  local dir = cwd
  while dir and dir ~= "" and dir ~= "/" do
    -- Check for .jj/repo/store/git (jj-native repo backed by git)
    local jj_git_config = dir .. "/.jj/repo/store/git/config"
    local owner, repo = parse_git_config(jj_git_config)
    if owner and repo then
      return owner, repo
    end

    -- Check for .jj/repo/store/git as a file (contains path to git dir)
    local jj_git_pointer = dir .. "/.jj/repo/store/git"
    local pf = io.open(jj_git_pointer, "r")
    if pf then
      local pointer_content = pf:read("*a")
      pf:close()
      -- If it's a path to a git directory, try reading its config
      local git_dir = pointer_content:gsub("%s+$", "")
      if git_dir ~= "" and not git_dir:find("^%[") then
        owner, repo = parse_git_config(git_dir .. "/config")
        if owner and repo then
          return owner, repo
        end
      end
    end

    -- Check for .git/config
    local git_config = dir .. "/.git/config"
    owner, repo = parse_git_config(git_config)
    if owner and repo then
      return owner, repo
    end

    -- Check for .git as a file (worktree)
    local git_file = dir .. "/.git"
    local gf = io.open(git_file, "r")
    if gf then
      local git_content = gf:read("*a")
      gf:close()
      local gitdir = git_content:match("gitdir:%s*(.+)")
      if gitdir then
        gitdir = gitdir:gsub("%s+$", "")
        if not gitdir:match("^/") then
          gitdir = dir .. "/" .. gitdir
        end
        owner, repo = parse_git_config(gitdir .. "/config")
        if owner and repo then
          return owner, repo
        end
      end
    end

    -- Move up one directory
    dir = dir:match("(.+)/[^/]+$")
  end

  return nil, nil
end

--- Get owner/repo for the current working directory.
--- Results are cached per cwd.
---@param cwd string|nil Optional cwd override; defaults to vim.fn.getcwd()
---@return string|nil owner
---@return string|nil repo
function M.detect(cwd)
  cwd = cwd or vim.fn.getcwd()

  -- Return cached result
  if cache[cwd] ~= nil then
    if cache[cwd] == false then
      return nil, nil
    end
    return cache[cwd].owner, cache[cwd].repo
  end

  local owner, repo = detect_from_dir(cwd)
  if owner and repo then
    cache[cwd] = { owner = owner, repo = repo }
  else
    cache[cwd] = false
  end

  return owner, repo
end

--- Build the API path prefix for the current repo.
--- Returns nil if repo cannot be detected.
---@param cwd string|nil
---@return string|nil path like "/api/repos/owner/repo"
function M.api_prefix(cwd)
  local owner, repo = M.detect(cwd)
  if not owner or not repo then
    return nil
  end
  return "/api/repos/" .. owner .. "/" .. repo
end

--- Clear the detection cache (useful when switching projects).
function M.clear_cache()
  cache = {}
end

--- Get a display string for the current repo.
---@param cwd string|nil
---@return string like "owner/repo" or "unknown"
function M.display(cwd)
  local owner, repo = M.detect(cwd)
  if owner and repo then
    return owner .. "/" .. repo
  end
  return "unknown"
end

return M
